metadata/metadata.go

```diff
@@ -0,0 +1,74 @@
+package metadata
+
+import (
+	"errors"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+var (
+	// These endpoints don't fit into the regexp below, but should still be
+	// allowed.
+	discoveryEndpoints = []string{
+		"",
+		"/",
+		"/0.1",
+		"/0.1/",
+		"/computemetadata",
+		"/computemetadata/",
+	}
+	// Explanation of regexp below:
+	//   - Lead with '/', and follow with either "0.1/meta-data",
+	//     "computeMetadata/v1beta1", or "computeMetadata/v1";
+	//   - Allow zero or more well-defined URL children that may only
+	//     contain alphanumerics, '_', or '-', and may include
+	//     '.' if it comes after '@';
+	//   - Finish with an optional '/'.
+	allowedPattern = regexp.MustCompile("^\\/(0\\.1\\/meta\\-data|computemetadata\\/v1beta1|computemetadata\\/v1)(\\/[0-9a-z_\\-]+(@[0-9a-z_\\-\\.]*)?)*\\/?$")
+	// Explanation of regexp below:
+	//   - Similar to above, lead with '/', and follow with either
+	//     "0.1/meta-data", "computeMetadata/v1beta1/instance", or
+	//     "computeMetadata/v1/instance";
+	//   - Follow with '/', then either "attributes/kube-env" or
+	//     "service-accounts/<valid-service-account-or-alias>/identity".
+	// In particular, note that '.' may only come after '@' in the valid
+	// service account or alias.
+	concealedPattern = regexp.MustCompile("^\\/(0\\.1\\/meta\\-data|computemetadata\\/v1beta1\\/instance|computemetadata\\/v1\\/instance)/(attributes/kube-env|service-accounts/[0-9a-z_\\-]+(@[0-9a-z_\\-\\.]*)?/identity)$")
```

Comment:
these are pretty painful. would it simplify things to clean the path first?

https://golang.org/pkg/path/#Clean

Reply:
Done.